### ISSUE TITLE
[rllib] local_mode=True in ray.init() causes tensorflow error

### ISSUE THREAD

**Issue by thomasbbrunner** [ray-project/ray#18421]
### What is the problem?

Enabling local_mode causes `TypeError: Variable is unhashable. Instead, use tensor.ref() as the key.` error to happen when initializing Keras Layers.

Versions:
ray: tested on 2.0.0.dev0 and 1.6.0
tf: 2.6.0
python: 3.9.5
os: ubuntu 21.04

### Reproduction

The error happens when creating the `base` variable in the `ActorCritic` class (see code below).

If `local_mode=False`, then the `base` layer is created successfully.
If `local_mode=True`, then tensorflow throws "TypeError: Variable is unhashable. Instead, use tensor.ref() as the key." error when creating the `base` layer.

The inputs to the constructor are exactly the same for both cases.

The only difference I could spot was in the shape of the `inputs` variable:
`local_mode=True` has `inputs = <KerasTensor: shape=(None, 3) dtype=float32 (created by layer 'input_1')>`
`local_mode=False` has `inputs = <KerasTensor: shape=(?, 3) dtype=float32 (created by layer 'input_1')>`

Notice the difference in the shapes? (None vs ?) Both should be valid shapes in Tensorflow. I tested this on a standalone tensorflow script (no ray) and the shape created was `(None, 3)` and it was successfully used to generate the `base` layer.

```python
import pdb
import tensorflow as tf
import ray
import ray.rllib
from ray.rllib.agents.impala import ImpalaTrainer
from ray.rllib.models.tf.tf_modelv2 import TFModelV2
from ray.rllib.models import ModelCatalog

class ActorCritic(TFModelV2):

    def __init__(self, obs_space, action_space, num_outputs, model_config, name):
        super().__init__(
            obs_space=obs_space, action_space=action_space,
            num_outputs=num_outputs, model_config=model_config,
            name=name)

        inputs = tf.keras.layers.Input(shape=obs_space.shape)
        pdb.set_trace() # debug
        base = tf.keras.layers.Dense(128, activation=tf.nn.relu)(inputs)
        actor = tf.keras.layers.Dense(num_outputs)(base)
        critic = tf.keras.layers.Dense(1)(base)
        self.model = tf.keras.Model(inputs=inputs, outputs=[actor, critic])
        self.critic_out = None

    def forward(self, input_dict, state, seq_lens):
        actor_out, self.critic_out = self.model(input_dict["obs_flat"])
        return actor_out, []

    def value_function(self):
        return self.critic_out


if __name__ == "__main__":

    ray.init(
        local_mode=True, # TODO toggle this!
    )

    ModelCatalog.register_custom_model(
        "actor_critic", ActorCritic)

    config = {
        "env": "Pendulum-v0",
        "model": {
            "custom_model": "actor_critic",
            "custom_model_config": {},},
        # enable eager execution
        "framework": "tfe",
        "log_level": "INFO",
    }

    trainer = ImpalaTrainer(config=config)
```

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by xwjiang2010** [comment]
Hi @thomasbbrunner,
Thanks for reporting this!
I was able to reproduce the same issue. It seems with local_mode=False, tf2.enabled() returns False while it returns True when  local_mode=True. This leads to the described behavior discrepancy. After reading more, it seems variable.__hash__() is no longer available starting Tensorflow 2.0, which leads to the unhashable error message when we try to put the ResourceVariable into a set. 
@sven1977 Does RLlib do anything special about TensorFlow 2.0 for local_mode v.s. not?
cc @gjoliver

**Comment by thomasbbrunner** [comment]
Ok, so I figured out what the problem is: I am importing tensorflow with `import tensorflow as tf` and not the recommended way using `try_import_tf()`. Somehow this leads to the described bug, probably because this causes the function `try_import_tf` to skip the following lines:
```python
tf1_module.disable_v2_behavior()
tf1_module.enable_resource_variables()
```
